drivers/mdio/mdio_xilinx_axienet.c

```diff
@@ -0,0 +1,200 @@
+
+#include <sys/types.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/mdio.h>
+#include <zephyr/sys/barrier.h>
+
+#define XILINX_AXIENET_MDIO_SETUP_REG_OFFSET                  0x00000500
+#define XILINX_AXIENET_MDIO_SETUP_REG_MDIO_DISABLE_MASK       0x0
+#define XILINX_AXIENET_MDIO_SETUP_REG_MDIO_ENABLE_MASK        (1 << 6)
+#define XILINX_AXIENET_MDIO_SETUP_REG_MDIO_CLOCK_DIVIDER_MASK ((1 << 6) - 1)
+#define XILINX_AXIENET_MDIO_CONTROL_REG_OFFSET                0x00000504
+#define XILINX_AXIENET_MDIO_CONTROL_REG_MASK_READY            (1 << 7)
+#define XILINX_AXIENET_MDIO_CONTROL_REG_SHIFT_PHYADDR         24
+#define XILINX_AXIENET_MDIO_CONTROL_REG_SHIFT_REGADDR         16
+#define XILINX_AXIENET_MDIO_CONTROL_REG_SHIFT_TXOP            14
+#define XILINX_AXIENET_MDIO_CONTROL_REG_MASK_READ             0x00008000
+#define XILINX_AXIENET_MDIO_CONTROL_REG_MASK_WRITE            0x00004000
+#define XILINX_AXIENET_MDIO_CONTROL_REG_MASK_INITIATE         (1 << 11)
+#define XILINX_AXIENET_MDIO_WRITE_DATA_REG_OFFSET             0x00000508
+#define XILINX_AXIENET_MDIO_READ_DATA_REG_OFFSET              0x0000050C
+#define XILINX_AXIENET_MDIO_READ_DATA_REG_DATA_MASK           0x0000ffff
+
+struct mdio_xilinx_axienet_config {
+	void *reg;
+	unsigned int clock_frequency_hz;
+};
+
+struct mdio_xilinx_axienet_data {
+	uint16_t clock_divider;
+	bool bus_enabled;
+};
+
+static void xilinx_axienet_mdio_write_register(const struct mdio_xilinx_axienet_config *config,
+					       uintptr_t reg_offset, uint32_t value)
+{
+	volatile uint32_t *reg_addr = (uint32_t *)((uint8_t *)(config->reg) + reg_offset);
+	*reg_addr = value;
+	barrier_dmem_fence_full(); /* make sure that write commits */
+}
+
+static uint32_t xilinx_axienet_read_mdio_register(const struct mdio_xilinx_axienet_config *config,
+						  uintptr_t reg_offset)
+{
+	const volatile uint32_t *reg_addr = (uint32_t *)((uint8_t *)(config->reg) + reg_offset);
+	const uint32_t ret = *reg_addr;
+
+	barrier_dmem_fence_full(); /* make sure that read commits */
+	return ret;
+}
+
+static void mdio_xilinx_axienet_bus_disable(const struct device *dev)
+{
+	const struct mdio_xilinx_axienet_config *config = dev->config;
+
+	LOG_INF("Disable MDIO Bus!");
+
+	xilinx_axienet_mdio_write_register(config, XILINX_AXIENET_MDIO_SETUP_REG_OFFSET,
+					   XILINX_AXIENET_MDIO_SETUP_REG_MDIO_DISABLE_MASK);
+}
+
+#define MDIO_TARGET_FREQUENCY_HZ 2500000 /* 2.5 MHz, i.e., max MDIO clock according to IEEE spec */
+#define MDIO_BUSY_WAIT_USEC      1
+#define MDIO_BUSY_WAIT_MAX_USEC  10000
+
+static void mdio_bus_wait_ready(const struct mdio_xilinx_axienet_config *config)
+{
+	int attempts = 0;
+
+	while ((xilinx_axienet_read_mdio_register(config, XILINX_AXIENET_MDIO_CONTROL_REG_OFFSET) &
+		XILINX_AXIENET_MDIO_CONTROL_REG_MASK_READY) == 0) {
+
+		LOG_DBG("Waiting for MDIO to become ready!");
+		k_busy_wait(MDIO_BUSY_WAIT_USEC);
+		attempts += MDIO_BUSY_WAIT_USEC;
+		if (attempts >= MDIO_BUSY_WAIT_MAX_USEC) {
```

Comment:
if this is really going to take this long, it could block up to 10 ms? that's a rather long time, could potentially cause some bad timing issues due to most of the phy drivers will have a periodic task to poll the phy status, is there an interrupt based approach that could be used instead? not blocking over this since I don't know the hardware but pointing it out is all, on some other mdio driver we had a complaint if it blocked for more than 1 ms

Reply:
The Xilinx MDIO does have an interrupt. I will test the interrupt-driven approach and get back to you.